iothub_client/src/iothub_message.c

```diff
@@ -213,57 +265,36 @@ IOTHUB_MESSAGE_HANDLE IoTHubMessage_Clone(IOTHUB_MESSAGE_HANDLE iotHubMessageHan
         }
         else
         {
-            result->messageId = NULL;
-            result->correlationId = NULL;
-            result->userDefinedContentType = NULL;
-            result->contentEncoding = NULL;
+            InitializeIoTHubMessage(result);
 
             if (source->messageId != NULL && mallocAndStrcpy_s(&result->messageId, source->messageId) != 0)
             {
                 LogError("unable to Copy messageId");
-                free(result);
+                IoTHubMessage_Destroy(result);
                 result = NULL;
             }
             else if (source->correlationId != NULL && mallocAndStrcpy_s(&result->correlationId, source->correlationId) != 0)
             {
                 LogError("unable to Copy correlationId");
-                if (result->messageId != NULL)
-                {
-                    free(result->messageId);
-                }
-                free(result);
+                IoTHubMessage_Destroy(result);
                 result = NULL;
             }
             else if (source->userDefinedContentType != NULL && mallocAndStrcpy_s(&result->userDefinedContentType, source->userDefinedContentType) != 0)
             {
                 LogError("unable to copy contentType");
-                if (result->messageId != NULL)
-                {
-                    free(result->messageId);
-                }
-                if (result->correlationId != NULL)
-                {
-                    free(result->correlationId);
-                }
-                free(result);
+                IoTHubMessage_Destroy(result);
                 result = NULL;
             }
             else if (source->contentEncoding != NULL && mallocAndStrcpy_s(&result->contentEncoding, source->contentEncoding) != 0)
             {
                 LogError("unable to copy contentEncoding");
-                if (result->messageId != NULL)
-                {
-                    free(result->messageId);
-                }
-                if (result->correlationId != NULL)
-                {
-                    free(result->correlationId);
-                }
-                if (result->userDefinedContentType != NULL)
-                {
-                    free(result->userDefinedContentType);
-                }
-                free(result);
+                IoTHubMessage_Destroy(result);
+                result = NULL;
+            }
+            else if (source->diagnosticId != NULL && mallocAndStrcpy_s(&result->diagnosticId, source->diagnosticId) != 0)
+            {
+                LogError("unable to copy diagnosticId");
+                IoTHubMessage_Destroy(result);
```

Comment:
Mixed feelings about this...
Thanks for the optimization, it makes it cleaner.
However, this change is not part of the diagnostics, and I rather not see them in this commit.
And usually we don't call public APIs within the code (at least we should not).

Reply:
ok, let's refactoring the code in separate commit.